[GUI][SubtitlesDialog] Add context menu options for subtitle addons

enen92 opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
The subtitle dialog (the dialog you use to download subtitles) lack context menu items on the list of available subtitle services. One of the things that annoys me the most after a fresh kodi installation, regarding subtitles, is the fact that I constantly forget that some of the services (such as opensubtitles) requires authentication. To be able to fill the credentials, and if you've just started playing some media content, you'd have to close the dialog, close the subtitle settings dialog, move back to the Kodi main window, get to the addon browser, find the subtitle addon in question and then configure it...and afterwards, repeat the playback process. We should be able to do it rigth from the subtitles dialog.

On the other hand, subtitle services are also prone to change (and get broken) over time. You usually only notice this when you attempt to download subtitles after playback has started. So, it'd be great if we could also disable those addons right from the same dialog.

That's what this PR does. It adds context menu items to the list of services displayed on the subtitles dialog, enabling 2 items: "Settings" and "Disable"

@CastagnaIT can you take a look?

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Improve user experience, mostly mine :)

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Runtime tested

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
They should be able to access subtitle addon settings right from the subtitle dialog. They can also disable subtitle addons if for some reason they don't work anymore.

## Screenshots (if appropriate):

![image](https://user-images.githubusercontent.com/7375276/135729150-fadaffdc-bb48-431e-829f-f49f3a58a23b.png)


## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [x] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

enen92 committed:
[GUI][SubtitlesDialog] Add context menu options for subtitle addons

ksooo commented on xbmc/video/dialogs/GUIDialogSubtitles.h:
```diff
@@ -50,6 +50,12 @@ class CGUIDialogSubtitles : public CGUIDialog, CJobQueue
+   * Called when the context menu is requested on a subtitle service
+   * present on the list of installed subtitle addons
+   * */
+  void OnSubtitleServiceContextMenu(const int itemIdx);
```
`const`makes no sense here.

enen92 replied:
Why so if the value is supposed to never be changed?

ksooo replied:
There is zero benefit in using `const` for by-value parameters, but there are downsides. If you later want to change the *implementation* that it shall change the value of the parameter, because you want to avoid introduction of another local var just to copy the const paramter value to be able to manipulate it, you have to change the header and to recompile everything dependent. Hassle without any need.

enen92 replied:
Well while your comment makes great sense as a generic rule of thumb I don't see how it can have much relevance in this particular context. The likelihood of having another dialog extending the subtitlesdialog class is pretty much inexistent. Furthermore this refers to the index of the item (on a list) which the context menu has just been requested, it is not supposed to be changed by the implementation (nor I foresee any plausible reason for it in the future). It's much more likely that someone, for some reason, wants to pass a const listitem reference.
If you agree, I'd prefer to keep it as is for now.

ksooo replied:
> If you agree, I'd prefer to keep it as is for now.

 Well, I do not really agree.

 Which value has the information for the reader of the header, that the copy (!) of the parameter's value they pass to the method will not be modified by the implementation of the method? We're not talking about references here, which would be a completely different story. We are talking about values that are copied onto the stack, then passed to the method. `const`is pointless here because the called method cannot manipulate the data the caller passes to the method.

enen92 replied:
I see what you mean now, since this will result in a copy of the original value it's pointless being declared const as we shouldn't care what the implementation does with it. I'll just remove `const` tomorrow as I don't think it makes much sense to pass it as a const ref.   
Thanks for pointing it out

ksooo replied:
you are welcome. :-)

howie-f commented on xbmc/video/dialogs/GUIDialogSubtitles.cpp:
```diff
@@ -380,6 +402,51 @@ void CGUIDialogSubtitles::OnSearchComplete(const CFileItemList *items)
   SetInvalid();
 }
 
+void CGUIDialogSubtitles::OnSubtitleServiceContextMenu(const int itemIdx)
```
it is the same (unwritten?) rule that applies to return values: never return from/pass in a plain const. it‘s either 1. `T` like here because it‘s an `int` not a heavy object, or 2. `T&`, or 3. `const T&`

ksooo replied:
Right.

CastagnaIT commented:
i have never used download subtitles services, but a first test context menus seem to works
if i choose `Browse for subtitle...` instead of `Download subtitle...` kodi crash brutally
but this is another thing

enen92 commented:
Thanks for testing Stefano. Not related to this PR but I've never experienced a crash on "Browse subtitles". Can you provide a stack trace? Which directory are you trying to browse? Some remote fs?

enen92 committed:
[GUIDialogSubtitles] Move gui defines to anonymous namespace

enen92 commented:
- Added a new commit that moves the old GUI defines to an anonymous namespace
- gui context button enum also moved to the new anonymous namespace
- drop the const modifier on the "callback" member

Should be good now

howie-f reviewed (approved):
:+1: LGTM

garbear commented on xbmc/video/dialogs/GUIDialogSubtitles.cpp:
```diff
@@ -380,6 +405,54 @@ void CGUIDialogSubtitles::OnSearchComplete(const CFileItemList *items)
+              g_localizeStrings.Get(24021));
+
+
+  auto idx = static_cast<SUBTITLE_SERVICE_CONTEXT_BUTTONS>(CGUIDialogContextMenu::Show(buttons));
```
Can you reduce the two-newlines to one?

enen92 replied:
done, thanks
